### ISSUE TITLE
Post-scan image processing

### ISSUE THREAD

**Issue by Mogaba** [sjamesr/jfreesane#28]
I'm not sure if this is an issue but I don't know where to ask this.

I'm trying to do some post-scan image processing: adjust brightness, contrast etc. The problem is that when setting scan mode to "Color" the image returned by **device.acquireImage()** has type 	**BufferedImage.TYPE_CUSTOM**. When writing the image to the file and reading it back from the file the image type becomes **BufferedImage.TYPE_3BYTE_BGR**. And RGB-values of the original image (of type TYPE_CUSTOM) are a bit larger than of file's image so the original image looks more brighter than it should be.

Could you please guide me how to convert the original TYPE_CUSTOM image to RGB-type without write-to/read-from file?

**Comment by sjamesr** [comment]
Hi @Mogaba, sorry for the delayed response. It might help me if you post expected and actual images demonstrating the change in brightness. In the meantime I'll see what I can do.

**Comment by sjamesr** [comment]
@Mogaba also, if you're still interested, I found a bug in JFreeSane's handling of 16 bit per pixel images. https://github.com/sjamesr/jfreesane/commit/ba323aee63115efa99e7909a2382cfe7afdbb441 should fix it, would you mind patching it in and see if it gives you better results when you write and then read the image file?

**Comment by Mogaba** [comment]
Hello @sjamesr , here's the question I asked at stackoverflow:
http://stackoverflow.com/questions/37478424/bufferedimage-color-saturation

And after some search I found this question:
http://stackoverflow.com/questions/12524623/what-are-the-practical-differences-when-working-with-colors-in-a-linear-vs-a-no

I took the formula from the answer for sRGB -> Linear transformation and applied it to each RGB-value of the original image and it worked - the image brightness turned to normal.

> I found a bug in JFreeSane's handling of 16 bit per pixel images.

Unfortunately the patch didn't work. How do you tell if you're working with 16- or 24-bit images? Every time I invoke **device.acquireImage()** it returns an image of type **BufferedImage.TYPE_CUSTOM**. But somehow after writing to a file with **ImageIO.write()** it becomes normal 24-bit RGB.

**Comment by Mogaba** [comment]
Here's the code I used for writing/reading:

```
BufferedImage bimg = dev.acquireImage();             // bad brightness
ImageIO.write(bimg, "png", new File("test.png"));
bimg = ImageIO.read(new File("test.png"));           // good brightness
```


And here's the code I used for sRGB -> Linear transformation;

```
BufferedImage bimg = dev.acquireImage();        // bad brightness
BufferedImage newBimg = new BufferedImage(bimg.getWidth(), bimg.getHeight(), BufferedImage.TYPE_INT_RGB);
newBimg.getGraphics().drawImage(bimg, 0, 0, null);
toLinear(newBimg);       //good brightness

public static void toLinear(BufferedImage bimg) {
		Color c;
		float[] newRGB;
		for (int x = 0; x < bimg.getWidth(); x++)
			for (int y = 0; y < bimg.getHeight(); y++) {
				c = new Color(bimg.getRGB(x, y));
				newRGB = convertPixel(new float[] { c.getRed() / 255f, c.getGreen() / 255f, c.getBlue() / 255f });
				bimg.setRGB(x, y,
						new Color(Math.round(newRGB[0] * 255), Math.round(newRGB[1] * 255), Math.round(newRGB[2] * 255))
								.getRGB());
			}
	}

public static float[] convertPixel(float[] in) {
		float[] out = new float[3];
		for (int i = 0; i < 3; i++) {
			float n = in[i];
			if (n < 0)
				n = 0f;
			if (n > 1)
				n = 1f;
			if (n <= 0.03928f)
				out[i] = (float) (n / 12.92);
			else
				out[i] = (float) (Math.exp(2.4 * Math.log((n + 0.055) / 1.055)));
		}
		return out;
	}
```

**Comment by sjamesr** [comment]
How are you verifying that the brightness is bad before you convert it to PNG?

**Comment by sjamesr** [comment]
I can reproduce this now, looking into it...

**Comment by sjamesr** [comment]
Can you try checking out the `use_8bits_always` branch and rebuilding your code using that? I think it fixes your particular problem, but the code will throw away the lower 8 bits of a 16-bit sample if you're using 16-bit sample depth.

**Comment by sjamesr** [comment]
Actually, a simpler fix appears to be to just use CS_sRGB instead of CS_LINEAR_RGB, please take a look at https://github.com/sjamesr/jfreesane/pull/40/commits/c47cb4121ad2517ec056d661d274c31d4c31aa39 and try it out.

**Comment by Mogaba** [comment]
I applied c47cb41 and it solved the problem, the brightness is ok. Thanks a lot!

**Comment by sjamesr** [comment]
Fixed in 0.96!
